### PR TITLE
Replaced the random generator from Pharo.

### PR DESCRIPTION
**Pull request by Mathilde411**
The current random generator of Pharo is heavily skewed in certain situations (for example the first value generated from a new seed is constrained in a certain range).

So this replaces the current generator with a new one based on Permuted Congruential Generators (https://www.pcg-random.org/index.html)

This uses a primitive implemented in pharo-project/pharo-vm#645

### CODE REVIEW

File: src/Random-Tests/RandomTest.class.st [hunk 1]
```diff
@@ -97,6 +85,205 @@ RandomTest >> testNextInto [
...
+RandomTest >> testPrimitiveRandomGeneration1 [
+
+	| state results |
+	results := OrderedCollection new.
+
+	state := DoubleWordArray new: 1.
+	state at: 1 put: 16r4d595df4d0f33173.
+
+	1 to: 100 do: [ :_ |
+	results add: (Random primitiveRandomNumber: state) ].
```
Comment (guillep) [inline src/Random-Tests/RandomTest.class.st]: Could we not have tests on private implementation-specific methods?

File: src/Random-Tests/RandomTest.class.st [hunk 2]
```diff
@@ -97,6 +85,205 @@ RandomTest >> testNextInto [
...
+RandomTest >> testPrimitiveRandomGeneration2 [
+
+	| state results |
+	results := OrderedCollection new.
+
+	state := DoubleWordArray new: 1.
+	state at: 1 put: 16r977afd8015414a94.
+
+	1 to: 100 do: [ :_ |
+	results add: (Random primitiveRandomNumber: state) ].
```
Comment (guillep) [inline src/Random-Tests/RandomTest.class.st]: same here

File: src/Random-Tests/RandomTest.class.st [hunk 3]
```diff
@@ -97,6 +85,205 @@ RandomTest >> testNextInto [
...
+RandomTest >> testPrimitiveRandomGeneration3 [
+
+	| state results |
+	results := OrderedCollection new.
+
+	state := DoubleWordArray new: 1.
+	state at: 1 put: 16re81fbe472905b453.
+
+	1 to: 100 do: [ :_ |
+	results add: (Random primitiveRandomNumber: state) ].
```
Comment (guillep) [inline src/Random-Tests/RandomTest.class.st]: and here

File: src/Random-Tests/RandomTest.class.st [hunk 4]
```diff
@@ -97,6 +85,205 @@ RandomTest >> testNextInto [
...
+RandomTest >> testPrimitiveRandomGeneration4 [
+
+	| state results |
+	results := OrderedCollection new.
+
+	state := DoubleWordArray new: 1.
+	state at: 1 put: 16re8adc5779db42e9b.
+
+	1 to: 100 do: [ :_ |
+	results add: (Random primitiveRandomNumber: state) ].
```
Comment (guillep) [inline src/Random-Tests/RandomTest.class.st]: and

File: src/Random-Tests/RandomTest.class.st [hunk 5]
```diff
@@ -97,6 +85,205 @@ RandomTest >> testNextInto [
...
+RandomTest >> testPrimitiveRandomGeneration5 [
+
+	| state results |
+	results := OrderedCollection new.
+
+	state := DoubleWordArray new: 1.
+	state at: 1 put: 16r39352fc390c369b2.
+
+	1 to: 100 do: [ :_ |
+	results add: (Random primitiveRandomNumber: state) ].
```
Comment (guillep) [inline src/Random-Tests/RandomTest.class.st]: here.

Even all these tests look similar excepting for some values, maybe some common code could be extracted into common methods?

File: src/Random-Tests/RandomTest.class.st [hunk 6]
```diff
@@ -97,6 +85,205 @@ RandomTest >> testNextInto [
...
+
+	
+	
+]
+
+{ #category : #tests }
+RandomTest >> testSeedStateGeneration2 [
+	
+	gen seed: 42.
+	self assert: (gen state at: 1) equals: 16r977afd8015414a94
```
Comment (guillep) [inline src/Random-Tests/RandomTest.class.st]: Why testing the state and not the next number only?

**Review by guillep — CHANGES_REQUESTED** [review]
Thanks @BastouP411, I left some comments

**Comment by Ducasse** [comment]
Can we have a nice class comment?
With a link to the exact implementation.

**Comment by Ducasse** [comment]
In the class comment I would like also to have the following code 

```
// *Really* minimal PCG32 code / (c) 2014 M.E. O'Neill / pcg-random.org
// Licensed under Apache License 2.0 (NO WARRANTY, etc. see website)

typedef struct { uint64_t state;  uint64_t inc; } pcg32_random_t;

uint32_t pcg32_random_r(pcg32_random_t* rng)
{
    uint64_t oldstate = rng->state;
    // Advance internal state
    rng->state = oldstate * 6364136223846793005ULL + (rng->inc|1);
    // Calculate output function (XSH RR), uses old state for max ILP
    uint32_t xorshifted = ((oldstate >> 18u) ^ oldstate) >> 27u;
    uint32_t rot = oldstate >> 59u;
    return (xorshifted >> rot) | (xorshifted << ((-rot) & 31));
}
```

and 

@techreport{oneill:pcg2014,
    title = "PCG: A Family of Simple Fast Space-Efficient Statistically Good Algorithms for Random Number Generation",
    author = "Melissa E. O'Neill",
    institution = "Harvey Mudd College",
    address = "Claremont, CA",
    number = "HMC-CS-2014-0905",
    year = "2014",
    month = Sep,
    xurl = "https://www.cs.hmc.edu/tr/hmc-cs-2014-0905.pdf",
}

Note that I found strange that it did not get published since 2014. I could not find a 

https://lemire.me/blog/2017/08/22/testing-non-cryptographic-random-number-generators-my-results/

**Comment by Mathilde411** [comment]
> @TechReport{oneill:pcg2014,
> title = "PCG: A Family of Simple Fast Space-Efficient Statistically Good Algorithms for Random Number Generation",
> author = "Melissa E. O'Neill",
> institution = "Harvey Mudd College",
> address = "Claremont, CA",
> number = "HMC-CS-2014-0905",
> year = "2014",
> month = Sep,
> xurl = "https://www.cs.hmc.edu/tr/hmc-cs-2014-0905.pdf",
> }

Done :)

**Comment by svenvc** [comment]
Correction for one link: https://www.pcg-random.org

I also like Wikipedia links: https://en.wikipedia.org/wiki/Permuted_congruential_generator

<blockquote><div>PCG, A Better Random Number Generator</div><div><strong><a href="http://www.pcg-random.org/index.html">PCG, A Family of Better Random Number Generators</a></strong></div><div>PCG is a family of simple fast space-efficient statistically good algorithms for random number generation.  Unlike many general-purpose RNGs, they are also hard to predict.</div></blockquote>
<blockquote><div><strong><a href="https://en.wikipedia.org/wiki/Permuted_congruential_generator">Permuted congruential generator - Wikipedia</a></strong></div></blockquote>

**Comment by svenvc** [comment]
I am no specialist at all, and of course the implementation is an engineering tradeoff, but this critique seems rather harsh: https://pcg.di.unimi.it/pcg.php


<blockquote><div><strong><a href="https://pcg.di.unimi.it/pcg.php">The wrap-up on PCG generators</a></strong></div></blockquote>

**Comment by svenvc** [comment]
> I am no specialist at all, and of course the implementation is an engineering tradeoff, but this critique seems rather harsh: https://pcg.di.unimi.it/pcg.php
> 
> > **[The wrap-up on PCG generators](https://pcg.di.unimi.it/pcg.php)**

Maybe it is all just noise, author's response to the critique: https://www.pcg-random.org/posts/on-vignas-pcg-critique.html

<blockquote><div><strong><a href="https://pcg.di.unimi.it/pcg.php">The wrap-up on PCG generators</a></strong></div></blockquote>
<blockquote><div>PCG, A Better Random Number Generator</div><div><strong><a href="http://www.pcg-random.org/posts/on-vignas-pcg-critique.html">On Vigna's PCG Critique</a></strong></div><div>On 14 May 2018, Sebastiano Vigna added a page to his website (archived here) entitled “The wrap-up on PCG generators” that attempts to persuade readers to avoid various PCG generators.
That day, he al</div></blockquote>

**Comment by guillep** [comment]
I see that at the end this uses no primitive right?

Could you publish here a screenshot on how the new generator behaves in the case of the original issue?

Also, there seem to be a couple of failing tests, are they related?

imagen

**Comment by Ducasse** [comment]
tx sven for the two links. This was interesting (especially the answer).

**Comment by Mathilde411** [comment]
> I see that at the end this uses no primitive right?

The primitive has not yet been implemented in the VM, the pull request is still out there.

> Could you publish here a screenshot on how the new generator behaves in the case of the original issue?

When running this:
```st
r := Random new.
((1 to: 10000) collect: [ :s | r seed: s. r next ]) rsHistogram 
```
We get something like this in the old random generator:
broken_random

The first generated from a new seed is highly biased (We only get numbers up to 0.08, and the limit is linear based on the seed.)

With the new generator, we do not have this bias:
repaired_random

**Comment by Mathilde411** [comment]
Now only the Roassal tests fail, but they are unimportant and will be fixed shortly ( see https://github.com/ObjectProfile/Roassal3/pull/569 ) !

**Pull request merged**